### PR TITLE
poplog findhelp --category=help --exact <topic>, closing #130

### LINKED ISSUE #130: Add findhelp subcommand to assist with VS Code integration
The next step for VS Code integration is to be able to retrieve Poplog documentation. However the Poplog help system is not declarative but procedural and not static but extensible. So an accurate integration requires an "API" into Poplog. The natural place to supply that API is through the poplog-command itself.

This change makes it possible to extend the poplog-command using autoloadable Pop-11 files, to provide a subcommand extension that can interrogate the file-based documentation using the internal library functions, and to list the results as JSON. 

In essence: 
- A command of the form `poplog findhelp <options> <topic>` that is available on a standard installation of GetPoplog. Options that take arguments will be of the form `--{option}={value}`. 
- Which allows for the specification of a single category e.g. `poplog findhelp --category=help <topic>`.
- There are 5 categories: help, teach, ref, doc, showlib. The default category is 'help'.
- An exact-match can be specified e.g. `poplog findhelp --exact <topic>`.
- The response will be a ranked list of matches in JSON format; a schema will be provided. It will include:
    - **version** of poplog

- Each match will include:
    - **quality** of match (from 0.0 to 1.0, 1.0 being best)
    - **title** of matching resource.
    - **summary** - a short description, which may not be available, signified by null.
    - **category** - n.b. a search in one category may legitimately yield matches in another)
    - **file-path** - the absolute file path as a string.
    - **from** line number in the file at which the description starts, may be null.
    - **to** line number in the file at which the description finishes, may be null.

The schema for that JSON is [available as a gist](https://gist.github.com/sfkleach/4065ef69d393c659e297420ebc8c0248) and also included below for convenience:
```json
{
  "$id": "https://github.com/GetPoplog/findhelp.schema.json",
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "title": "FindHelp",
  "type": "object",
  "properties": {
    "version": {
      "description": "A string describing the version of Poplog.",
      "type": "string"
    },
    "documentation": {
      "description": "Matching documentation.",
      "type": "array",
      "items": {
        "type": "object",
        "properties": {
          "quality": {
            "description": "A number from 0 to 1.0 indicating the match quality, with 1 being a perfect match.",
            "type": "number"
          },
          "category": {
            "description": "The category of documentation, reflecting the level formality and technical detail.",
            "type": "string",
            "enum": [ "help", "teach", "doc", "ref" ]
          },
          "title": {
            "description": "The title of the resource.",
            "type": "string"
          },
          "summary": {
            "description": "An optional summary of the resource, which may be null.",
            "type": ["string", "null"]
          },
          "path": {
            "description": "File path of the resource.",
            "type": "string"
          },
          "from": {
            "description": "Line number the description starts on (1-indexed), which may be null.",
            "type": ["integer", "null"]
          },
          "to": {
            "description": "Line number of the last line of the description (inclusive), which may be null.",
            "type": ["integer", "null"]
          }
        },
        "required": ["quality", "title", "summary", "path", "from", "to", "category"]
      }
    },
    "required": ["popversion", "documentation"]
  }
}
```

### PR DESCRIPTION
**Pull request by sfkleach**
This change makes it possible to extend the poplog-command tool using autoloadable Pop-11 files, provides a subcommand extension that can interrogate the file-based documentation using the internal library functions, and to list the results as JSON. See #130.

There are several parts to this PR:
 - Some pop-11 functions for providing the search functionality and JSON translation; these are added to the `$usepop/pop/getpoploglib/` folder.
 - A modification of the `poplog` command tool so that new subcommands can be added.
 - A new `findhelp` subcommand.

## Notes
1. The pop-11 support code includes an implementation of named-records, which I have called "dict"s in deliberate imitation of Python's dicts. I regard this implementation as the first version of an important missing feature. It also includes implementations of `discinline` and `incharline` that are imports from the GOSPL library. Over time I want to integrate all of GOSPL into GetPoplog.
2. All the extensions that are specific to the GetPoplog distribution are placed in the `getpoploglib/` folder. This folder is only linked into Poplog's search path if a programmer includes `uses getpoplog` in their project.
3. The `poplog` command tool is a C-program that is generated at build-time. At the moment it is effectively a template that is implemented as a Bash script. Fortunately this change was uncomplicated. As the script gets more complicated we should probably move the generation to another tool. Python is the GetPoplog scripting tool of choice.

### CODE REVIEW

**Comment by sfkleach** [comment]
Firstly, I note that `poplog findhelp` is not documented in the `--help` option. It plainly needs to be documented but I rather feel the best place would be under `$usepop/pop/getpoploglib/help/getpoplog_subcommand_findhelp`. However the VSC work depends on this, so this fix will come later. UPDATE: See issue #138.

**Comment by sfkleach** [comment]
Secondly, I am concerned that not all the variables of the `dict` library will autoload e.g. `isdict`. That's a mistake and needs rectifying. However, it too can wait for a follow-up, since this work is on the critical path. UPDATE: See issue #137.

**Comment by sfkleach** [comment]
Thirdly, the documentation for `json_print`, `json_println` and `json_pprint` is missing. This will be backfilled. UPDATE: See is #136.

**Comment by sfkleach** [comment]
What about working with other Poplog distributions? It would make sense if, as a fallback, we either (1) provide a webservice that performs this search or (2) compile a static database that can be downloaded, e.g. SQLITE, along with a procedure for searching it. Both solutions suffer from the serious flaw that they cannot include locally installed packages. However, a webservice would be quite easy to provide (although file-paths would need to be replaced by URLs).

File: makePoplogCommander.sh [hunk 1]
```diff
@@ -938,12 +938,14 @@ cat << \****
...
-        fprintf( stderr, "\n" );
-        return EXIT_FAILURE;
+        setUpEnvironment( base, flags, envv );
+        char * subpath = "/pop/getpoploglib/lib/getpoplog_run_subcommand.p";
+        char * path = safe_malloc( strlen( subpath ) + strlen( base ) + 1 );
+        char * p = stpcpy( path, base );
+        p = stpcpy( p, subpath );
+        deque_push_front( argd, path );
+        deque_push_front( argd, "pop11" );
+        execvp( "pop11", deque_as_array( argd ) );
```
Comment (willprice) [inline makePoplogCommander.sh:949]: Should we not interrogate the return value and display an error to the user? Or is the idea that running pop11 will display that error message?

Shouldn't we exit with a non-zero exit code if this call returns a non-zero exit code?

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p [hunk 2]
```diff
@@ -0,0 +1,62 @@
+compile_mode :pop11 +strict;
+
+section;
+
+uses getpoplog;
+
+;;; Given command line options of the form
```
Comment (willprice) [inline base/pop/getpoploglib/lib/getpoplog_run_subcommand.p:7]: nit: seems a bit odd that short args with a single dash would end up in the second list.

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            endif 
+        endfor;
+    %];
+    return( newdict_from_twinlists(keys, values), others )
+enddefine;
+
+define getpoplog_run_subcommand();
+    if poparglist.null then
+        dlocal cucharout = cucharerr;
+        npr( 'Missing options for poplog command' );
```
Comment (willprice) [inline base/pop/getpoploglib/lib/getpoplog_run_subcommand.p]:
```suggestion
        npr( 'Expected name of subcommand but none was given' );
```

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        false -> pop_exit_ok;
+        sysexit();
+    else
+        lvars (subcmd, args) = poparglist.dest;
+        lvars w = consword( 'getpoplog_subcommand_' <> subcmd );
+        if sys_autoload( w ) then
+            valof( w )( parse_args( args ) )
+        else
+            dlocal cucharout = cucharerr;
+            pr( 'Unexpected arguments: ' );
```
Comment (willprice) [inline base/pop/getpoploglib/lib/getpoplog_run_subcommand.p]: Is it an unexpected argument or an unknown subcommand?

File: base/pop/getpoploglib/other/findhelp.schema.json [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  "$schema": "https://json-schema.org/draft/2020-12/schema",
+  "title": "FindHelp",
+  "type": "object",
+  "properties": {
+    "popversion": {
+      "description": "A string describing the version of Poplog.",
+      "type": "string"
+    },
+    "documentation": {
+      "description": "Matching documentation.",
```
Comment (willprice) [inline base/pop/getpoploglib/other/findhelp.schema.json:12]: Matching what?

File: base/pop/getpoploglib/other/findhelp.schema.json [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+      "type": "array",
+      "items": {
+        "type": "object",
+        "properties": {
+          "quality": {
+            "description": "A number from 0 to 1.0 indicating the match quality, with 1 being a perfect match.",
+            "type": "number"
+          },
+          "category": {
+            "description": "The category of documentation, reflecting the level formality and technical detail.",
```
Comment (willprice) [inline base/pop/getpoploglib/other/findhelp.schema.json:22]:
```suggestion
            "description": "The category of documentation, reflecting the level of formality and technical detail.",
```

File: base/pop/lib/lib/getpoplog.p [hunk 7]
```diff
@@ -3,6 +3,6 @@
 ;;;     - section independent
 ;;;     - safe against multiple reloads
 
-extend_searchlist( '$usepop/pop/getpoploglibs/auto', popautolist ) -> popautolist;
-extend_searchlist( '$usepop/pop/getpoploglibs/lib', popliblist ) -> popliblist;
-extend_searchlist( '$usepop/pop/getpoploglibs/help', vedhelplist ) -> vedhelplist;
+extend_searchlist( '$usepop/pop/getpoploglib/auto', popautolist ) -> popautolist;
```
Comment (willprice) [inline base/pop/lib/lib/getpoplog.p:6]: Was this a bug fix, or did we just not have a `getpoploglibs` folder before?

File: base/pop/lib/lib/getpoplog.p [hunk 8]
```diff
@@ -3,6 +3,6 @@
 ;;;     - section independent
 ;;;     - safe against multiple reloads
 
-extend_searchlist( '$usepop/pop/getpoploglibs/auto', popautolist ) -> popautolist;
-extend_searchlist( '$usepop/pop/getpoploglibs/lib', popliblist ) -> popliblist;
```
Comment (willprice) [inline base/pop/lib/lib/getpoplog.p:7]: Was this a bug? Why the change from popliblist to popuseslist?

File: makePoplogCommander.sh [hunk 9]
```diff
@@ -938,12 +938,14 @@ cat << \****
...
     }
     perror( NULL );
     return EXIT_FAILURE;
```
Comment (willprice) [inline makePoplogCommander.sh:953]: Doesn't this mean we'll always exit with a failed return status even when the subcommand was successful?

File: makePoplogCommander.sh (hunk 1 above)
Comment (sfkleach) [reply makePoplogCommander.sh:949]: Pop11 will return a suitable exit code & error message. I will add a comment to that effect. UPDATE: edited makePoplogCommander.sh & the change will appear on the next push.

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p (hunk 2 above)
Comment (sfkleach) [reply base/pop/getpoploglib/lib/getpoplog_run_subcommand.p:7]: Oh so sharp-eyed! Yes, this is a pretty crude first attempt at processing command line options. It needs more work and that is why it isn't exposed as a library as yet. Another infelicity in the same vein is that there is no way to turn off option processing.

I will raise a suitable issue. UPDATE: Issue #135 raised to include this tech-debt.

File: base/pop/getpoploglib/other/findhelp.schema.json (hunk 5 above)
Comment (sfkleach) [reply base/pop/getpoploglib/other/findhelp.schema.json:12]: Yes good point! Fixed in the gist. Sadly gists cannot be made organisational & I couldn't think of a sensible place to include this file in the git archive.

File: base/pop/lib/lib/getpoplog.p (hunk 7 above)
Comment (sfkleach) [reply base/pop/lib/lib/getpoplog.p:6]: Bug fix!

File: base/pop/lib/lib/getpoplog.p (hunk 8 above)
Comment (sfkleach) [reply base/pop/lib/lib/getpoplog.p:7]: It was a bug.

File: makePoplogCommander.sh (hunk 9 above)
Comment (sfkleach) [reply makePoplogCommander.sh:953]: The execvp will only return on a failure - otherwise the running process is replaced by the process running Poplog. I will add a comment. UPDATE: Comment added in my local branch and it will appear in the next push.

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p (hunk 3 above)
Comment (sfkleach) [reply base/pop/getpoploglib/lib/getpoplog_run_subcommand.p]: Agreed. I have made the change in my local branch and it will appear on the next push.

File: base/pop/getpoploglib/lib/getpoplog_run_subcommand.p (hunk 4 above)
Comment (sfkleach) [reply base/pop/getpoploglib/lib/getpoplog_run_subcommand.p]: The latter - good point. Message fixed in local branch - will appear in next push.

File: base/pop/getpoploglib/other/findhelp.schema.json (hunk 6 above)
Comment (sfkleach) [reply base/pop/getpoploglib/other/findhelp.schema.json:22]: Nice! Fixed in gist.

File: makePoplogCommander.sh (hunk 9 above)
Comment (willprice) [reply makePoplogCommander.sh:953]: Really? I didn't get that understanding when reading the man page: https://man7.org/linux/man-pages/man3/perror.3.html
are you 100% sure?

```c
#include <errno.h>
#include <unistd.h>
char* argv[];
int main()  {
execvp( "blah", argv);
perror(NULL);
return 0;
}
```
running this returns 0.

**Review by willprice — CHANGES_REQUESTED** [review]

File: makePoplogCommander.sh (hunk 9 above)
Comment (sfkleach) [reply makePoplogCommander.sh:953]: The perror should only be reached if the execvp fails to find "blah". From the online manual:

> The exec() functions return only if an error has occurred.  The return value is -1, and [errno](https://man7.org/linux/man-pages/man3/errno.3.html) is set to indicate the error.

There's no need to check the return value because it is guaranteed to be -1.

**Comment by willprice** [comment]
Ah, `execvp` doesn't return if the command runs successfully, I thought you were referring to `perror` when you were saying this earlier. Got it, lgtm :)

**Review by willprice — APPROVED** [review]
